checker/cost.go

```diff
@@ -0,0 +1,432 @@
+	target := call.GetTarget()
+	args := call.GetArgs()
+
+	var sum CostEstimate
+
+	argTypes := make([]AstNode, len(args))
+	for i, arg := range args {
+		// TODO: && || short circuit, so min cost should only include 1st arg eval
+		// unless exhaustive evaluation is enabled
+		sum = sum.Add(c.cost(arg))
+		argTypes[i] = expr{t: c.getType(arg), expr: arg}
+	}
+
+	ref := c.checker.ReferenceMap[e.Id]
+	if ref == nil || len(ref.GetOverloadId()) == 0 {
+		return CostEstimate{}
+	}
+	var targetType AstNode
+	if target != nil {
+		if call.Target != nil {
+			sum = sum.Add(c.cost(call.Target))
+			targetType = expr{t: c.getType(call.Target), expr: call.Target}
+		}
+	}
+	// Pick a cost estimate range that covers all the overload cost estimation ranges
+	fnCost := CostEstimate{Min: uint64(math.MaxUint64), Max: 0}
+	for _, overload := range ref.GetOverloadId() {
+		overloadCost := c.functionCost(overload, &targetType, argTypes)
+		if overloadCost.Max > fnCost.Max {
+			fnCost.Max = overloadCost.Max
+		}
+		if overloadCost.Min < fnCost.Min {
+			fnCost.Min = overloadCost.Min
+		}
+	}
+	return sum.Add(fnCost)
+}
+
+func (c *coster) costCreateList(e *exprpb.Expr) CostEstimate {
+	create := e.GetListExpr()
+	var sum CostEstimate
+	for _, e := range create.Elements {
+		sum = sum.Add(c.cost(e))
+	}
+	return sum.Add(createListBaseCost)
+}
+
+func (c *coster) costCreateStruct(e *exprpb.Expr) CostEstimate {
+	str := e.GetStructExpr()
+	if str.MessageName != "" {
+		return c.costCreateMessage(e)
+	} else {
+		return c.costCreateMap(e)
+	}
+}
+
+func (c *coster) costCreateMap(e *exprpb.Expr) CostEstimate {
+	mapVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range mapVal.GetEntries() {
+		key := ent.GetMapKey()
+		sum = sum.Add(c.cost(key))
+
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMapBaseCost)
+}
+
+func (c *coster) costCreateMessage(e *exprpb.Expr) CostEstimate {
+	msgVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range msgVal.GetEntries() {
+		sum = sum.Add(c.cost(ent.Value))
+	}
+	return sum.Add(createMessageBaseCost)
+}
+
+func (c *coster) costComprehension(e *exprpb.Expr) CostEstimate {
+	comp := e.GetComprehensionExpr()
+	var sum CostEstimate
+	sum = sum.Add(c.cost(comp.IterRange))
+	sum = sum.Add(c.cost(comp.AccuInit))
+	accuType := c.getType(comp.AccuInit)
+	rangeType := c.getType(comp.IterRange)
+	var varType *exprpb.Type
+
+	switch kindOf(rangeType) {
+	case kindList:
+		varType = rangeType.GetListType().ElemType
+	case kindMap:
+		// Ranges over the keys.
+		varType = rangeType.GetMapType().KeyType
+	case kindDyn, kindError, kindTypeParam:
+		// Set the range iteration variable to type DYN as well.
+		varType = decls.Dyn
+	default:
+		// not a valid range
+		varType = decls.Error
+	}
+
+	// Create a scope for the comprehension since it has a local accumulation variable.
+	// This scope will contain the accumulation variable used to compute the result.
+	c.env = c.env.enterScope()
```

Comment:
With the `type_map` change, I think you can remove the dependence on the `*Env` and the scope management used below.

Reply:
For field paths I still need to do some scope tracking but I was able to drop all the type computations.